### PR TITLE
Receive charge prices

### PR DESCRIPTION
**Pull request by Mech0z**
## Description
As a Charge owner
I want to submit charge prices

**Acceptance criteria:**

- [ ] 1 Given a Charge owner submits a charge price series that passes sync validation
When DataHub has processed the charge price series
Then the owner is able to peek a confirmation message
- [ ] 2 Process flow for prices added to GitHub
- [ ] 3 Architecture diagram for prices added to GitHub

## References

* #1288

### CODE REVIEW

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Handlers/ChargeAndPriceHandler.cs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        }
+
+        public async Task HandleAsync(ChargeCommandReceivedEvent commandReceivedEvent)
+        {
+            switch (commandReceivedEvent.Command.Document.BusinessReasonCode)
+            {
+                case BusinessReasonCode.UpdatePriceInformation:
+                    await _messageDispatcher.DispatchAsync(new ChargeCommandAcceptedEvent(
+                            commandReceivedEvent.PublishedTime,
+                            commandReceivedEvent.Command))
```
Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Handlers/ChargeAndPriceHandler.cs:44]: Use ChargeCommandAcceptedEventFactory?

**Review by prtandrup — COMMENTED** [review]
A few comments, still :eyes:

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Handlers/ChargeAndPriceHandler.cs (hunk 1 above)
Comment (Mech0z) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Handlers/ChargeAndPriceHandler.cs]: changed

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Handlers/IChargeAndPriceHandler.cs [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System.Threading.Tasks;
+using GreenEnergyHub.Charges.Domain.Dtos.ChargeCommandReceivedEvents;
+
+namespace GreenEnergyHub.Charges.Application.Charges.Handlers
+{
+    public interface IChargeAndPriceHandler
```
[the remark this answers is not in the document]
Comment (Mech0z) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Handlers/IChargeAndPriceHandler.cs:20]: renamed to ChargeHandler

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/MarketParticipants/BusinessReasonCode.cs [hunk 3]
```diff
@@ -22,5 +22,6 @@ public enum BusinessReasonCode
         Unknown = 0,
         UpdateMasterDataSettlement = 1,
         UpdateChargeInformation = 2,
+        UpdatePriceInformation = 3,
```
[the remark this answers is not in the document]
Comment (Mech0z) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/MarketParticipants/BusinessReasonCode.cs:25]: Renamed to UpdateChargePrices

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.MessageHub/Models/AvailableChargeData/AvailableChargeData.cs [hunk 4]
```diff
@@ -39,8 +39,8 @@ public class AvailableChargeData : AvailableDataBase
             Instant startDateTime,
             Instant endDateTime,
             VatClassification vatClassification,
-            bool taxIndicator,
-            bool transparentInvoicing,
+            TaxIndicator taxIndicator,
+            TransparentInvoicing transparentInvoicing,
```
Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.MessageHub/Models/AvailableChargeData/AvailableChargeData.cs:43]: I suggest you let the two props `TaxIndicator` and `TransparentInvoicing` remain booleans, and thus avoid doing a database update on the Message.AvailableChargeData Table which might have to be revoked later. 
Perhaps just handle the enum mappings in the `AvailableChargeDataFactory.cs`.

And as we are only aiming at providing confirmations for prices, I would add a check to either `ChargeDataAvailableNotifierEndpoint` or `AvailableDataNotifier` to filter out acceptedEvents with BusinessReasonCode = UpdateChargePrices, so nothing price series related is stored in the MessageHub table.

Comment (Mech0z) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.MessageHub/Models/AvailableChargeData/AvailableChargeData.cs]: reverted

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Application/Charges/Handlers/ChargeCommandHandlerTests.cs [hunk 5]
```diff
@@ -31,22 +32,26 @@ public class ChargeCommandHandlerTests
...
+        public async Task ChangeOfChargesTransactionHandler_WhenCalledWithCharge_ShouldCallChargePublisher(
+            [Frozen] Mock<IMessageDispatcher<ChargeCommandReceivedEvent>> chargeEventPublisher,
             ChargeCommandHandler sut)
         {
             // Arrange
-            var transaction = new ChargeCommandBuilder().Build();
+            var document = new DocumentDtoBuilder()
+                .WithBusinessReasonCode(BusinessReasonCode.UpdateChargeInformation)
+                .Build();
+            var transaction = new ChargeCommandBuilder().WithDocumentDto(document).Build();
```
Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Application/Charges/Handlers/ChargeCommandHandlerTests.cs:43]:
```suggestion
            var command = new ChargeCommandBuilder().WithDocumentDto(document).Build();
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Application/Charges/Handlers/ChargeCommandHandlerTests.cs [hunk 6]
```diff
@@ -31,22 +32,26 @@ public class ChargeCommandHandlerTests
     {
         [Theory]
         [InlineAutoDomainData]
-        public async Task ChangeOfChargesTransactionHandler_WhenCalled_ShouldCallPublisher(
-            [Frozen] Mock<IMessageDispatcher<ChargeCommandReceivedEvent>> localEventPublisher,
+        public async Task ChangeOfChargesTransactionHandler_WhenCalledWithCharge_ShouldCallChargePublisher(
```
Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Application/Charges/Handlers/ChargeCommandHandlerTests.cs:35]:
```suggestion
        public async Task HandleAsync_WhenCalledWithChargeCommand_ShouldDispatchReceivedEvent(
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Application/Charges/Handlers/ChargeHandlerTests.cs [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+using Xunit.Categories;
+
+namespace GreenEnergyHub.Charges.Tests.Application.Charges.Handlers
+{
+    [UnitTest]
+    public class ChargeHandlerTests
+    {
+        [Theory]
+        [InlineAutoDomainData]
+        public async Task HandleAsync_WhenBusinessReasonCodeIsUpdatePriceInformation_SendConfirmation(
```
Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Application/Charges/Handlers/ChargeHandlerTests.cs:35]:
```suggestion
        public async Task HandleAsync_WhenBusinessReasonCodeIsUpdateChargePrices_SendConfirmation(
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/MessageHub/Models/AvailableChargeData/AvailableChargeDataFactoryTests.cs [hunk 8]
```diff
@@ -19,8 +19,10 @@
 using FluentAssertions;
 using GreenEnergyHub.Charges.Application.Messaging;
 using GreenEnergyHub.Charges.Core.DateTime;
+using GreenEnergyHub.Charges.Domain.Charges;
 using GreenEnergyHub.Charges.Domain.Dtos.ChargeCommands;
 using GreenEnergyHub.Charges.Domain.MarketParticipants;
+using GreenEnergyHub.Charges.Infrastructure.Core.Cim.Charges;
```
Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/MessageHub/Models/AvailableChargeData/AvailableChargeDataFactoryTests.cs:25]: using can be removed again

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTest.Core/Fixtures/FunctionApp/ChargesServiceBusResourceNames.cs [hunk 9]
```diff
@@ -58,6 +58,10 @@ public static class ChargesServiceBusResourceNames
         public const string CommandRejectedTopicKey = "command-rejected";
         public const string CommandRejectedSubscriptionName = "command-rejected";
 
+        // Internal, price command, received
```
Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTest.Core/Fixtures/FunctionApp/ChargesServiceBusResourceNames.cs:61]: remove?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure.Core/Cim/Charges/TransparentInvoicingMapper.cs [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System.ComponentModel;
+using GreenEnergyHub.Charges.Domain.Charges;
+
+namespace GreenEnergyHub.Charges.Infrastructure.Core.Cim.Charges
+{
+    public class TransparentInvoicingMapper
```
[the remark this answers is not in the document]
Comment (Mech0z) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure.Core/Cim/Charges/TransparentInvoicingMapper.cs:20]: added tests

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTest.Core/Fixtures/FunctionApp/ChargesServiceBusResourceNames.cs (hunk 9 above)
Comment (Mech0z) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTest.Core/Fixtures/FunctionApp/ChargesServiceBusResourceNames.cs:61]: done

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/MessageHub/Models/AvailableChargeData/AvailableChargeDataFactoryTests.cs (hunk 8 above)
Comment (Mech0z) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/MessageHub/Models/AvailableChargeData/AvailableChargeDataFactoryTests.cs:25]: done

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Handlers/IChargeHandler.cs [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+namespace GreenEnergyHub.Charges.Application.Charges.Handlers
+{
+    /// <summary>
+    /// Delegates a charge depending on if it contains
+    /// Charge Prices or Charge information.
+    /// </summary>
+    public interface IChargeHandler
+    {
+        /// <summary>
+        /// Synchronously handle the event.
```
Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Charges/Handlers/IChargeHandler.cs:27]:
```suggestion
        /// Asynchronously handle the event.
```

**Review by prtandrup — APPROVED** [review]
LGTM. Suggest we remove the AC2 and AC3 in the PR description and take them separately.

**Pull request merged**